core: rework dependency system to be based on atoms + add three new dep types

poettering opened the pull request:
Some new addition for the dependency system, in order to address #13386, #9869, #4263

poettering commented:
(still lacks man page updates, hence the "wip" and "dont-merge thing for now)

dnicolodi commented on src/core/unit-dependency-atom.c:
```diff
@@ -0,0 +1,191 @@
+
+static const UnitDependencyAtom atom_map[_UNIT_DEPENDENCY_MAX] = {
+        /* A table that maps high-level dependency types to low-level dependency "atoms". The latter actually
+         * describe specific facets of dependency behaviour. The former combine them into one user-facing
```
Wouldn't it be clearer if the concept were to be called `UnitDependencyFacet`?

poettering replied:
well, naming things is always the hardest.

But I am pretty sure "atom" is the right name here, as we should emphasize that this is about splitting up dependency concepts into their basic, atomic building blocks. Hence "atom" sounds like the right word, since it nicely emphasizes that you can't split it up any further.

dnicolodi replied:
Indeed. Reading the patch Atom was not very informative and the whole thing made sense only when I arrived at the bottom of it where the concept is defined. If the hunks would have been in the reverse order in the commit, I would have probably not questioned the name :)

keszybz commented on src/core/unit-dependency-atom.c:
```diff
@@ -0,0 +1,191 @@
+        [UNIT_TRIGGERS]               = UNIT_ATOM_TRIGGERS,
+        [UNIT_TRIGGERED_BY]           = UNIT_ATOM_TRIGGERED_BY,
+        [UNIT_PROPAGATES_RELOAD_TO]   = UNIT_ATOM_PROPAGATES_RELOAD_TO,
+        [UNIT_RELOAD_PROPAGATED_FROM] = 0,
```
Hmm, so this would mean that the dependency is nought... This at least needs a comment.

FWIW, I think that it's not necessary to separate those single-line items out, or even comment about this. For the reader, it is obvious that those are single-bit (and one zero-bit) aliases just by looking.

poettering replied:
i'll move the zero-atom deps to their own section further down, with a comment about things.

keszybz commented on src/core/unit-dependency-atom.c:
```diff
@@ -0,0 +1,191 @@
+        [UNIT_REFERENCED_BY]          = UNIT_ATOM_REFERENCED_BY,
+};
+
+UnitDependencyAtom unit_dependency_to_atom(UnitDependency d) {
+        if (d < 0)
+                return _UNIT_DEPENDENCY_ATOM_INVALID;
+
+        assert(d < _UNIT_DEPENDENCY_MAX);
+
+        return atom_map[d];
+}
```
`static inline` ?

poettering replied:
dunno, i like encapsulation. and gcc LTO should figure this out on its own. i mean, "static inline" is pretty much identical to LTO stuff in effect — at least for internal APIs like this one.

keszybz commented on src/core/unit.h:
```diff
@@ -125,11 +135,13 @@ typedef struct Unit {
 
         Set *aliases; /* All the other names. */
 
-        /* For each dependency type we maintain a Hashmap whose key is the Unit* object, and the value encodes why the
-         * dependency exists, using the UnitDependencyInfo type */
-        Hashmap *dependencies[_UNIT_DEPENDENCY_MAX];
+        /* For each dependency type we can look up another Hashmap with this, whose key is a Unit* object,
+         * and whose value encodes why the dependency exists, using the UnitDependencyInfo type. i.e. a
+         * Hashmap(UnitDependency → Hashmap(Unit* → UnitDependencyInfo)) */
+        Hashmap *dependencies;
```
How does this affect memory usage in the typical scenario?

poettering replied:
I didn't measure it (it's not obvious to measure, due to the allocation cache). I'd expect moderate savings, since we no longer allocate space for per-dependency-type pointers to hashmap we almost never use. Given that sparely populated hashmaps are a single cache-backed allocations typically I think we basically make the Unit structure allocation quite a bit shorter with the Hashmap taking up much less space.

or to say this in math: we have now 33 dep types. on 64bit an array with a pointer for each is 264 bytes. A sparsely populated Hashmap costs only 48 bytes however (and as we have an allocation cache only one extra ptr for the cache maintainance, plus one pointer to the Hashmap from Unit). Hence 64 bytes with this PR vs. 264 bytes without it. Runtime behaviour should be slightly better with the PR as when iterating through the deps we only need to iterate through the dep types we actually have deps for instead of all dep types all the times.

(That is, if I didn#t make a mistake in my calculations)

keszybz commented on src/core/unit-dependency-atom.h:
```diff
@@ -0,0 +1,77 @@
+
+        /* This unit pulls in the other unit as JOB_START job into the transaction, and if that doesn't work
+         * the transaction fails. */
+        UNIT_ATOM_PULL_IN_MATTERS                     = UINT64_C(1) << 0,
+        /* Similar, but if it doesn't work, ignore. */
+        UNIT_ATOM_PULL_IN_DOESNT_MATTER               = UINT64_C(1) << 1,
+        /* Pull in a JOB_VERIFY job into the transaction, i.e. pull in JOB_VERIFY rather than
+         * JOB_START. i.e. check the unit is started but don't pull it in. */
+        UNIT_ATOM_PULL_IN_VERIFY                      = UINT64_C(1) << 2,
```
Maybe
UNIT_ATOM_PULL_IN_START                      (c.f. UNIT_ATOM_RETROACTIVE_START_… below)
UNIT_ATOM_PULL_IN_START_IGNORED
UNIT_ATOM_PULL_IN_VERIFY
UNIT_ATOM_STOP
UNIT_ATOM_STOP_IGNORED

keszybz commented on src/core/mount.c:
```diff
@@ -116,16 +134,33 @@ static bool mount_is_bind(const MountParameters *p) {
 }
 
-static bool mount_is_bound_to_device(const Mount *m) {
+static bool mount_is_bound_to_device(Mount *m) {
```
Losing `const` here is a bit unfortunate.
Maybe change get_mount_parameters() into a macro (MOUNT_PARAMETERS?). This will allow us to constify a bunch callers.

poettering replied:
Quite frankly I don't see the big benefit of const. i.e. macros are ugly beasts, and whether to have a const on some internal, static function somewhere sounds like a tiny benefit to me.

keszybz commented on test/units/testsuite-57.sh:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+set -ex
```
`-eux`

keszybz reviewed (commented):
Very interesting.

dtardon commented:
I think the PropagateStopTo=/StopPropagatedFrom= pair of dependency types obsoletes https://github.com/systemd/systemd/pull/11340.

poettering commented:
addressed the raised issues (except for the points i commented on, see above). Also added docs for the new dep types to the man page. dropping the WIP/dont-merge stuff hence. 

ptal

poettering commented:
Force pushed a new version addressing the typo @behrmann found. Hopefully this one should pass CI.

mrc0mmand commented:
@poettering after installing this revision on a ubuntu machine the system becomes completely unusable. Also, after reboot and some tinkering, this seems to be the culprit:

![image](https://user-images.githubusercontent.com/679338/116220743-7a7fbc00-a73c-11eb-8ba1-38cb0fac15d6.png)

mrc0mmand commented:
It is actually pretty easy to trigger:

```
# sysctl -w fs.suid_dumpable=1 kernel.core_pattern=/core.%p
fs.suid_dumpable = 1
kernel.core_pattern = /core.%p
# while systemctl daemon-reload; do :; done

Broadcast message from systemd-journald@ubuntu1804.localdomain (Tue 2021-04-27 11:13:04 UTC):

systemd[1]: Caught <ABRT>, dumped core as pid 1031.


Broadcast message from systemd-journald@ubuntu1804.localdomain (Tue 2021-04-27 11:13:04 UTC):

systemd[1]: Freezing execution.

Failed to reload daemon: Connection reset by peer
```

I'll get a symbolized stack trace shortly.

mrc0mmand commented:
Sort version:
```gdb
#0  0x00007f444ccfd2a7 in kill () at ../sysdeps/unix/syscall-template.S:78
#1  0x00005559d41dc97f in crash (sig=6) at ../src/core/main.c:256
#2  <signal handler called>
#3  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#4  0x00007f444ccfe921 in __GI_abort () at abort.c:79
#5  0x00007f444e588621 in log_assert_failed (text=0x7f444e6dfeb9 "e->key == i->next_key", file=0x7f444e6dfe1d "src/basic/hashmap.c", line=660, func=0x7f444e6e0220 <__PRETTY_FUNCTION__.8822> "hashmap_iterate_in_internal_order")
    at ../src/basic/log.c:860
#6  0x00007f444e579295 in hashmap_iterate_in_internal_order (h=0x5559d55fc908, i=0x7fffb31b8a70) at ../src/basic/hashmap.c:660
#7  0x00007f444e579369 in hashmap_iterate_entry (h=0x5559d55fc908, i=0x7fffb31b8a70) at ../src/basic/hashmap.c:704
#8  0x00007f444e579396 in _hashmap_iterate (h=0x5559d55fc908, i=0x7fffb31b8a70, value=0x7fffb31b8a60, key=0x7fffb31b8a68) at ../src/basic/hashmap.c:712
#9  0x00005559d432d791 in hashmap_iterate (h=0x5559d55fc908, i=0x7fffb31b8a70, value=0x7fffb31b8a60, key=0x7fffb31b8a68) at ../src/basic/hashmap.h:283
#10 0x00005559d432dc3c in target_add_default_dependencies (t=0x5559d55173d0) at ../src/core/target.c:50
#11 0x00005559d432dd79 in target_load (u=0x5559d55173d0) at ../src/core/target.c:77
#12 0x00005559d42bf4d8 in unit_load (u=0x5559d55173d0) at ../src/core/unit.c:1603
#13 0x00005559d42575f5 in manager_dispatch_load_queue (m=0x5559d550a510) at ../src/core/manager.c:1986
#14 0x00005559d4257a87 in manager_load_unit (m=0x5559d550a510, name=0x5559d5606070 "systemd-rfkill.socket", path=0x0, e=0x0, _ret=0x7fffb31b8be0) at ../src/core/manager.c:2116
#15 0x00005559d42c75bf in unit_add_dependency_by_name (u=0x5559d5606470, d=UNIT_WANTS, name=0x5559d5606070 "systemd-rfkill.socket", add_reference=true, mask=UNIT_DEPENDENCY_UDEV) at ../src/core/unit.c:3155
#16 0x00005559d4315b5f in device_add_udev_wants (u=0x5559d5606470, dev=0x5559d556f010) at ../src/core/device.c:406
#17 0x00005559d4316ad0 in device_setup_unit (m=0x5559d550a510, dev=0x5559d556f010, path=0x5559d550db20 "/sys/devices/virtual/misc/rfkill", main=true) at ../src/core/device.c:549
#18 0x00005559d4316c19 in device_process_new (m=0x5559d550a510, dev=0x5559d556f010) at ../src/core/device.c:580
#19 0x00005559d4318062 in device_enumerate (m=0x5559d550a510) at ../src/core/device.c:862
#20 0x00005559d42557f2 in manager_enumerate (m=0x5559d550a510) at ../src/core/manager.c:1533
#21 0x00005559d425fcdc in manager_reload (m=0x5559d550a510) at ../src/core/manager.c:3880
#22 0x00005559d41e514e in invoke_main_loop (m=0x5559d550a510, saved_rlimit_nofile=0x7fffb31b9220, saved_rlimit_memlock=0x7fffb31b9230, ret_reexecute=0x7fffb31b914b, ret_retval=0x7fffb31b9150, ret_shutdown_verb=0x7fffb31b91a0, 
    ret_fds=0x7fffb31b91b8, ret_switch_root_dir=0x7fffb31b9190, ret_switch_root_init=0x7fffb31b9198, ret_error_message=0x7fffb31b91a8) at ../src/core/main.c:1905
#23 0x00005559d41e8af4 in main (argc=1, argv=0x7fffb31b9428) at ../src/core/main.c:2892
```

<details>
<summary>Long version:</summary>

```gdb
#0  0x00007f444ccfd2a7 in kill () at ../sysdeps/unix/syscall-template.S:78
No locals.
#1  0x00005559d41dc97f in crash (sig=6) at ../src/core/main.c:256
        sa = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0 <repeats 16 times>}}, sa_flags = 0, sa_restorer = 0x0}
        pid = 1073
        __PRETTY_FUNCTION__ = "crash"
        __func__ = "crash"
#2  <signal handler called>
No locals.
#3  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
        set = {__val = {8927191084203854339, 0, 0, 0, 550421365844, 93844319833328, 140736198279168, 0, 11086856369975670938, 13763596107, 18446744073709551615, 0, 140736198313792, 139931348863744, 93845035417599, 18446744056529682435}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#4  0x00007f444ccfe921 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x1d55fd260, sa_sigaction = 0x1d55fd260}, sa_mask = {__val = {0, 139931350336032, 2835992783721, 139931350335005, 111669149696, 7868514768, 0, 140736198314176, 139931348927958, 0, 0, 0, 
              0, 93844319335376, 139931350343000, 139931350336032}}, sa_flags = 1315831325, sa_restorer = 0x7f444e6dfeb9}
        sigs = {__val = {32, 0 <repeats 15 times>}}
        __cnt = <optimized out>
        __set = <optimized out>
        __cnt = <optimized out>
        __set = <optimized out>
#5  0x00007f444e588621 in log_assert_failed (text=0x7f444e6dfeb9 "e->key == i->next_key", file=0x7f444e6dfe1d "src/basic/hashmap.c", line=660, func=0x7f444e6e0220 <__PRETTY_FUNCTION__.8822> "hashmap_iterate_in_internal_order")
    at ../src/basic/log.c:860
No locals.
#6  0x00007f444e579295 in hashmap_iterate_in_internal_order (h=0x5559d55fc908, i=0x7fffb31b8a70) at ../src/basic/hashmap.c:660
        e = 0x5559d555e180
        idx = 15
        __PRETTY_FUNCTION__ = "hashmap_iterate_in_internal_order"
#7  0x00007f444e579369 in hashmap_iterate_entry (h=0x5559d55fc908, i=0x7fffb31b8a70) at ../src/basic/hashmap.c:704
No locals.
#8  0x00007f444e579396 in _hashmap_iterate (h=0x5559d55fc908, i=0x7fffb31b8a70, value=0x7fffb31b8a60, key=0x7fffb31b8a68) at ../src/basic/hashmap.c:712
        e = 0xffffffff
        data = 0x0
        idx = 4294967295
#9  0x00005559d432d791 in hashmap_iterate (h=0x5559d55fc908, i=0x7fffb31b8a70, value=0x7fffb31b8a60, key=0x7fffb31b8a68) at ../src/basic/hashmap.h:283
No locals.
#10 0x00005559d432dc3c in target_add_default_dependencies (t=0x5559d55173d0) at ../src/core/target.c:50
        _dt = _UNIT_DEPENDENCY_INVALID
        _found = true
        __unique_prefix_data4 = {match_atom = UNIT_ATOM_ADD_DEFAULT_TARGET_DEPENDENCY_QUEUE, by_type = 0x5559d55fc908, by_unit = 0x5559d558e3d0, current_type = 0x2, by_type_iterator = {idx = 5, next_key = 0x8}, by_unit_iterator = {
            idx = 4294967295, next_key = 0x0}, current_unit = 0x7fffb31b89f8}
        other = 0x0
        r = 0
        __PRETTY_FUNCTION__ = "target_add_default_dependencies"
#11 0x00005559d432dd79 in target_load (u=0x5559d55173d0) at ../src/core/target.c:77
        t = 0x5559d55173d0
        r = 0
        __PRETTY_FUNCTION__ = "target_load"
#12 0x00005559d42bf4d8 in unit_load (u=0x5559d55173d0) at ../src/core/unit.c:1603
        r = 0
        __PRETTY_FUNCTION__ = "unit_load"
        __func__ = "unit_load"
#13 0x00005559d42575f5 in manager_dispatch_load_queue (m=0x5559d550a510) at ../src/core/manager.c:1986
        u = 0x5559d55173d0
        n = 141
        __PRETTY_FUNCTION__ = "manager_dispatch_load_queue"
#14 0x00005559d4257a87 in manager_load_unit (m=0x5559d550a510, name=0x5559d5606070 "systemd-rfkill.socket", path=0x0, e=0x0, _ret=0x7fffb31b8be0) at ../src/core/manager.c:2116
        r = 0
        __PRETTY_FUNCTION__ = "manager_load_unit"
---Type <return> to continue, or q <return> to quit---#15 0x00005559d42c75bf in unit_add_dependency_by_name (u=0x5559d5606470, d=UNIT_WANTS, name=0x5559d5606070 "systemd-rfkill.socket", add_reference=true, mask=UNIT_DEPENDENCY_UDEV) at ../src/core/unit.c:3155
        buf = 0x0
        other = 0x5559d55823a0
        r = 0
        __PRETTY_FUNCTION__ = "unit_add_dependency_by_name"
#16 0x00005559d4315b5f in device_add_udev_wants (u=0x5559d5606470, dev=0x5559d556f010) at ../src/core/device.c:406
        word = 0x5559d56279b0 "systemd-rfkill.socket"
        k = 0x5559d5606070 "systemd-rfkill.socket"
        added = 0x0
        wants = 0x0
        property = 0x5559d437c98c "SYSTEMD_WANTS"
        d = 0x5559d5606470
        r = 0
        __PRETTY_FUNCTION__ = "device_add_udev_wants"
        __func__ = "device_add_udev_wants"
#17 0x00005559d4316ad0 in device_setup_unit (m=0x5559d550a510, dev=0x5559d556f010, path=0x5559d550db20 "/sys/devices/virtual/misc/rfkill", main=true) at ../src/core/device.c:549
        e = 0x5559d55a4bf0 "sys-devices-virtual-misc-rfkill.device"
        sysfs = 0x5559d550db20 "/sys/devices/virtual/misc/rfkill"
        u = 0x5559d5606470
        delete = true
        r = 0
        __PRETTY_FUNCTION__ = "device_setup_unit"
        __func__ = "device_setup_unit"
#18 0x00005559d4316c19 in device_process_new (m=0x5559d550a510, dev=0x5559d556f010) at ../src/core/device.c:580
        sysfs = 0x5559d550db20 "/sys/devices/virtual/misc/rfkill"
        dn = 0x5559d5512518 "@\025wND\177"
        alias = 0x7fffb31b8e50 "\200\216\033\263\377\177"
        devnum = 139931348870681
        r = -716102376
        __PRETTY_FUNCTION__ = "device_process_new"
        __func__ = "device_process_new"
#19 0x00005559d4318062 in device_enumerate (m=0x5559d550a510) at ../src/core/device.c:862
        sysfs = 0x5559d5673010 "/sys/devices/pnp0/00:00/tty/ttyS0"
        e = 0x5559d55d43a0
        dev = 0x5559d556f010
        r = 1
        __PRETTY_FUNCTION__ = "device_enumerate"
        __func__ = "device_enumerate"
#20 0x00005559d42557f2 in manager_enumerate (m=0x5559d550a510) at ../src/core/manager.c:1533
        c = UNIT_DEVICE
        __PRETTY_FUNCTION__ = "manager_enumerate"
        __func__ = "manager_enumerate"
#21 0x00005559d425fcdc in manager_reload (m=0x5559d550a510) at ../src/core/manager.c:3880
        reloading = 0x0
        fds = 0x5559d56011e0
        f = 0x5559d55a6260
        r = 0
        __PRETTY_FUNCTION__ = "manager_reload"
        __func__ = "manager_reload"
#22 0x00005559d41e514e in invoke_main_loop (m=0x5559d550a510, saved_rlimit_nofile=0x7fffb31b9220, saved_rlimit_memlock=0x7fffb31b9230, ret_reexecute=0x7fffb31b914b, ret_retval=0x7fffb31b9150, ret_shutdown_verb=0x7fffb31b91a0, 
    ret_fds=0x7fffb31b91b8, ret_switch_root_dir=0x7fffb31b9190, ret_switch_root_init=0x7fffb31b9198, ret_error_message=0x7fffb31b91a8) at ../src/core/main.c:1905
        saved_log_target = _LOG_TARGET_INVALID
        saved_log_level = -1
        r = 2
        __PRETTY_FUNCTION__ = "invoke_main_loop"
        __func__ = "invoke_main_loop"
#23 0x00005559d41e8af4 in main (argc=1, argv=0x7fffb31b9428) at ../src/core/main.c:2892
        initrd_timestamp = {realtime = 0, monotonic = 0}
        userspace_timestamp = {realtime = 1619522913623512, monotonic = 1946443}
        kernel_timestamp = {realtime = 1619522911677069, monotonic = 0}
        security_start_timestamp = {realtime = 1619522913625248, monotonic = 1948179}
---Type <return> to continue, or q <return> to quit---        security_finish_timestamp = {realtime = 1619522913625959, monotonic = 1948890}
        saved_rlimit_nofile = {rlim_cur = 1024, rlim_max = 4096}
        saved_rlimit_memlock = {rlim_cur = 65536, rlim_max = 65536}
        skip_setup = false
        loaded_policy = false
        queue_default_job = true
        first_boot = false
        reexecute = false
        switch_root_dir = 0x0
        switch_root_init = 0x0
        before_startup = 1990039
        after_startup = 2135834
        systemd = "systemd"
        timespan = "pQ\232ND\177\000\000\336\303\323LD\177\000\000\300\204P\325YU\000\000\220\202P\325YU\000\000\001\000\000\000\000\000\000\000\335\034\063\324YU\000\000@\253xND\177\000\000\000\000\000\000\000\000\000"
        shutdown_verb = 0x0
        error_message = 0x0
        r = 0
        retval = 1
        m = 0x5559d550a510
        fds = 0x0
        __PRETTY_FUNCTION__ = "main"
        __func__ = "main"
```
</details>

Hopefully it'll help a bit.

poettering commented:
@mrc0mmand hmm, i tried hard reproducing this on fedora in both qemu and nspawn, but no chance. This appears to be ubuntu specific?

poettering commented:
@mrc0mmand so i tried to reproduce this on ubuntu too now, via mkosi built images. couldn't reproduce either. it's rock solid here, to run daemon-reload in a tight loop. the rfkill device that appears to trigger this in the backtrace you posted shows up fine too. any idea wht i could try?

mrc0mmand commented:
Interesting, let me try if I can reproduce it again or if it magically fixed itself.

mrc0mmand commented:
@poettering so far I haven't be able to reproduce it, could you try rebasing and force-pushing this PR to re-trigger CIs?

poettering commented:
> @poettering so far I haven't be able to reproduce it, could you try rebasing and force-pushing this PR to re-trigger CIs?

yes, done!

mrc0mmand commented:
Hmpf, it still fails as it did before... I'll take a closer look tomorrow, today I wasn't able to reproduce it on Fedora Rawhide nor on CentOS 8.

mrc0mmand commented:
I failed to reproduce it on Fedora Rawhide, CentOS 8.3, RHEL 8.3, Ubuntu 18.04, and Ubuntu 20.10. However, looking at the Ubuntu CI artifacts the issue remains the same:

```
May 12 21:34:36 autopkgtest systemd[1]: systemd-logind.service: Attaching egress BPF program to cgroup /sys/fs/cgroup/unified/system.slice/systemd-logind.service failed: Operation not permitted
May 12 21:34:36 autopkgtest systemd[1]: Reloading D-Bus System Message Bus.
May 12 21:34:36 autopkgtest dbus-daemon[396]: Unknown username "whoopsie" in message bus configuration file
May 12 21:34:36 autopkgtest dbus-daemon[396]: [system] Reloaded configuration
May 12 21:34:36 autopkgtest dbus-send[11079]: method return time=1620855276.920203 sender=org.freedesktop.DBus -> destination=:1.53 serial=3 reply_serial=2
May 12 21:34:36 autopkgtest systemd[1]: Reloaded D-Bus System Message Bus.
May 12 21:34:36 autopkgtest systemd-udevd[1309]: /lib/udev/rules.d/50-udev-default.rules:42 Unknown group 'sgx', ignoring
May 12 21:34:36 autopkgtest systemd-udevd[1309]: /lib/udev/rules.d/84-nm-drivers.rules:10 Invalid value "/bin/sh -c 'ethtool -i $1 | sed -n s/^driver:\ //p' -- $env{INTERFACE}" for PROGRAM (char 24: invalid substitution type), ignoring, but please fix it.
May 12 21:34:37 autopkgtest systemd[1]: Reloading.
May 12 21:34:37 autopkgtest systemd-sysv-generator[11106]: SysV service '/etc/init.d/apport' lacks a native systemd unit file. Automatically generating a unit file for compatibility. Please update package to include a native systemd unit file, in order to make it more safe and robust.
May 12 21:34:37 autopkgtest systemd-sysv-generator[11106]: SysV service '/etc/init.d/grub-common' lacks a native systemd unit file. Automatically generating a unit file for compatibility. Please update package to include a native systemd unit file, in order to make it more safe and robust.
May 12 21:34:37 autopkgtest systemd-sysv-generator[11106]: SysV service '/etc/init.d/rng-tools' lacks a native systemd unit file. Automatically generating a unit file for compatibility. Please update package to include a native systemd unit file, in order to make it more safe and robust.
May 12 21:34:37 autopkgtest systemd[1]: /lib/systemd/system/dbus.socket:5: ListenStream= references a path below legacy directory /var/run/, updating /var/run/dbus/system_bus_socket → /run/dbus/system_bus_socket; please update the unit file accordingly.
May 12 21:34:37 autopkgtest systemd[1]: Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:660, function hashmap_iterate_in_internal_order(). Aborting.
May 12 21:34:37 autopkgtest audit[11110]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 pid=11110 comm="systemd" exe="/lib/systemd/systemd" sig=6 res=1
May 12 21:34:37 autopkgtest systemd-coredump[11111]: Due to PID 1 having crashed coredump collection will now be turned off.
May 12 21:34:37 autopkgtest systemd[1]: Caught <ABRT>, dumped core as pid 11110.
May 12 21:34:37 autopkgtest systemd[1]: Freezing execution.
May 12 21:34:37 autopkgtest systemd-coredump[11111]: Process 11110 (systemd) of user 0 dumped core.
                                                     
                                                     Found module linux-vdso.so.1 with build-id: 2b7ed19b02338f4da08cb0560955a2d194211a4a
                                                     Found module libgpg-error.so.0 with build-id: f958b7b9ce43ccad1ea543b42cadfe8663f63705
                                                     Found module libuuid.so.1 with build-id: 4938b15a667a41cc98755418556ea492500a927a
                                                     Found module libattr.so.1 with build-id: 8685d51b54f3bf1718e1c306340b72bb713f6c59
                                                     Found module libcap-ng.so.0 with build-id: 7a1459569dc248f22fed84238124d34967d1aff4
                                                     Found module libpcre.so.3 with build-id: 5b3416bb188eaf3fa4b7530aae6c1890b38b0372
                                                     Found module libpthread.so.0 with build-id: 68f36706eb2e6eee4046c4fdca2a19540b2f6113
                                                     Found module libdl.so.2 with build-id: c5396cfcf808b3137c70c50f87a21bb9cebd4824
                                                     Found module liblzma.so.5 with build-id: 8fbcca354d964860b9e6eb3736e9b7bc6177b417
                                                     Found module libzstd.so.1 with build-id: cc33145236e295160b5ef00ed3da9c5e9a559e8b
                                                     Found module liblz4.so.1 with build-id: b532baa25dfb63a7f6e83d6dbb9dced180529ead
                                                     Found module libip4tc.so.0 with build-id: 71c2e2afc0336190ba1a88eba721a4c4a66a403d
                                                     Found module libgcrypt.so.20 with build-id: ace9a26e351aa92bd1c6e4a0a3946f2b15a0001c
                                                     Found module libcrypt.so.1 with build-id: 042e7dbe1099b30f45b45347d164e0417c10809d
                                                     Found module libcap.so.2 with build-id: 8adaca88bfc9825b942d9f6b55e733ad1829e855
                                                     Found module libblkid.so.1 with build-id: bbe20d3ad910e39003554ef3317a14ce834b57e7
                                                     Found module libacl.so.1 with build-id: 70129e819d8b316c2c8d1052a6f6dbe4b052774a
                                                     Found module ld-linux-x86-64.so.2 with build-id: 977c39fe87abfa426d3043f6c8e21f7be3f0e876
                                                     Found module libapparmor.so.1 with build-id: 4cd5101ad610ccc96e40176f9e069aed69ee42b1
                                                     Found module libkmod.so.2 with build-id: 431e488dae49e46e8a4dc53db237b5c7009a5a3b
                                                     Found module libaudit.so.1 with build-id: 6edf087d60b202690111bf0c612bce8c1cb3c0ec
                                                     Found module libpam.so.0 with build-id: 8421be6021d6b47b70ba053a295b34619b0a09f9
                                                     Found module libmount.so.1 with build-id: c3e565680f4f33fdfabce12db4b079ccab4700f1
                                                     Found module libselinux.so.1 with build-id: bc6660abcafb3cef8f7c775c6f848366b69f17fa
                                                     Found module libseccomp.so.2 with build-id: 3a290cfffcf4a580f1579a0fc9988346eb180264
                                                     Found module librt.so.1 with build-id: 8bb230e1d019753b019313b61a920224f1a372c9
                                                     Found module libsystemd-shared-248.so with build-id: d31e1ca8c3dcce790fc8dccaa48b3d49548a1b70
                                                     Found module libc.so.6 with build-id: ce450eb01a5e5acc7ce7b8c2633b02cc1093339e
                                                     Found module systemd with build-id: 5119eada4d13af3749e0b0b0e465085ccec9e6c6
                                                     Stack trace of thread 11110:
                                                     #0  0x00007f4a4cb202a7 kill (libc.so.6 + 0x3f2a7)
                                                     #1  0x0000559c08748240 n/a (systemd + 0x5c240)
                                                     #2  0x00007f4a4cb20040 n/a (libc.so.6 + 0x3f040)
                                                     #3  0x00007f4a4cb1ffb7 raise (libc.so.6 + 0x3efb7)
                                                     #4  0x00007f4a4cb21921 abort (libc.so.6 + 0x40921)
                                                     #5  0x00007f4a4c99a772 log_assert_failed (libsystemd-shared-248.so + 0x15b772)
                                                     #6  0x00007f4a4c940fff n/a (libsystemd-shared-248.so + 0x101fff)
                                                     #7  0x00007f4a4c943552 _hashmap_iterate (libsystemd-shared-248.so + 0x104552)
                                                     #8  0x0000559c08766a67 n/a (systemd + 0x7aa67)
                                                     #9  0x0000559c087d6f78 n/a (systemd + 0xeaf78)
                                                     #10 0x0000559c0877c495 n/a (systemd + 0x90495)
                                                     #11 0x0000559c087f3d82 n/a (systemd + 0x107d82)
                                                     #12 0x0000559c087f464c n/a (systemd + 0x10864c)
                                                     #13 0x0000559c087e3828 n/a (systemd + 0xf7828)
                                                     #14 0x0000559c087e3fcd n/a (systemd + 0xf7fcd)
                                                     #15 0x0000559c087e4ff7 n/a (systemd + 0xf8ff7)
                                                     #16 0x0000559c0877db25 n/a (systemd + 0x91b25)
                                                     #17 0x0000559c087457eb n/a (systemd + 0x597eb)
                                                     #18 0x0000559c0873baa9 n/a (systemd + 0x4faa9)
                                                     #19 0x00007f4a4cb02bf7 __libc_start_main (libc.so.6 + 0x21bf7)
                                                     #20 0x0000559c0873c6ba n/a (systemd + 0x506ba)
May 12 21:34:37 autopkgtest kernel: systemd: 2 output lines suppressed due to ratelimiting
May 12 21:35:02 autopkgtest dbus-daemon[396]: [system] Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
May 12 21:35:27 autopkgtest dbus-daemon[396]: [system] Failed to activate service 'org.freedesktop.systemd1': timed out (service_start_timeout=25000ms)
```

I'll deal with some other stuff first and then try to get into this further.

mrc0mmand commented:
@poettering ha, I think I got it - don't ask me why, but running `systemctl preset-all` just before the `while` loop does the trick:

<details>
<summary>Ubuntu 18.04</summary>

```
root@ubuntu1804:~# systemctl preset-all
Unit bootlogd.service is masked, ignoring.
Unit cryptdisks.service is masked, ignoring.
Unit bootlogs.service is masked, ignoring.
Unit mdadm.service is masked, ignoring.
Unit mdadm-waitidle.service is masked, ignoring.
Unit hostname.service is masked, ignoring.
Unit mountall.service is masked, ignoring.
Unit sendsigs.service is masked, ignoring.
Unit cryptdisks-early.service is masked, ignoring.
Unit hwclock.service is masked, ignoring.
Unit reboot.service is masked, ignoring.
Unit checkroot-bootclean.service is masked, ignoring.
Unit halt.service is masked, ignoring.
Unit mountnfs.service is masked, ignoring.
Unit killprocs.service is masked, ignoring.
Unit mountall-bootclean.service is masked, ignoring.
Unit rcS.service is masked, ignoring.
Unit rc.service is masked, ignoring.
Unit fuse.service is masked, ignoring.
Unit stop-bootlogd-single.service is masked, ignoring.
Unit mountnfs-bootclean.service is masked, ignoring.
Unit mountkernfs.service is masked, ignoring.
Unit lvm2.service is masked, ignoring.
Unit mountdevsubfs.service is masked, ignoring.
Unit sudo.service is masked, ignoring.
Unit stop-bootlogd.service is masked, ignoring.
Unit bootmisc.service is masked, ignoring.
Unit checkfs.service is masked, ignoring.
Unit rmnologin.service is masked, ignoring.
Unit single.service is masked, ignoring.
Unit umountfs.service is masked, ignoring.
Unit umountnfs.service is masked, ignoring.
Unit motd.service is masked, ignoring.
Unit screen-cleanup.service is masked, ignoring.
Unit checkroot.service is masked, ignoring.
Unit umountroot.service is masked, ignoring.
Unit x11-common.service is masked, ignoring.
Removed /etc/systemd/system/network-online.target.wants/systemd-networkd-wait-online.service.
Unit /lib/systemd/system/bootlogd.service is masked, ignoring.
Created symlink /etc/systemd/system/sysinit.target.wants/systemd-pstore.service → /lib/systemd/system/systemd-pstore.service.
Created symlink /etc/systemd/system/sockets.target.wants/systemd-userdbd.socket → /lib/systemd/system/systemd-userdbd.socket.
Unit /lib/systemd/system/cryptdisks.service is masked, ignoring.
Created symlink /etc/systemd/system/multi-user.target.wants/remote-veritysetup.target → /lib/systemd/system/remote-veritysetup.target.
Unit /lib/systemd/system/bootlogs.service is masked, ignoring.
Unit /lib/systemd/system/mdadm.service is masked, ignoring.
Unit /lib/systemd/system/mdadm-waitidle.service is masked, ignoring.
Unit /lib/systemd/system/hostname.service is masked, ignoring.
Unit /lib/systemd/system/mountall.service is masked, ignoring.
Unit /lib/systemd/system/sendsigs.service is masked, ignoring.
Unit /lib/systemd/system/cryptdisks-early.service is masked, ignoring.
Unit /lib/systemd/system/hwclock.service is masked, ignoring.
Unit /lib/systemd/system/reboot.service is masked, ignoring.
Unit /lib/systemd/system/checkroot-bootclean.service is masked, ignoring.
Unit /lib/systemd/system/halt.service is masked, ignoring.
Unit /lib/systemd/system/mountnfs.service is masked, ignoring.
Unit /lib/systemd/system/killprocs.service is masked, ignoring.
Created symlink /etc/systemd/system/ctrl-alt-del.target → /lib/systemd/system/reboot.target.
Unit /lib/systemd/system/mountall-bootclean.service is masked, ignoring.
Unit /lib/systemd/system/rcS.service is masked, ignoring.
Unit /lib/systemd/system/rc.service is masked, ignoring.
Created symlink /etc/systemd/system/multi-user.target.wants/machines.target → /lib/systemd/system/machines.target.
Unit /lib/systemd/system/fuse.service is masked, ignoring.
Unit /lib/systemd/system/stop-bootlogd-single.service is masked, ignoring.
Unit /lib/systemd/system/mountnfs-bootclean.service is masked, ignoring.
Unit /lib/systemd/system/mountkernfs.service is masked, ignoring.
Created symlink /etc/systemd/system/dbus-org.freedesktop.oom1.service → /lib/systemd/system/systemd-oomd.service.
Created symlink /etc/systemd/system/multi-user.target.wants/systemd-oomd.service → /lib/systemd/system/systemd-oomd.service.
Created symlink /etc/systemd/system/multi-user.target.wants/remote-cryptsetup.target → /lib/systemd/system/remote-cryptsetup.target.
Created symlink /etc/systemd/system/vmtoolsd.service → /lib/systemd/system/open-vm-tools.service.
Created symlink /etc/systemd/system/multi-user.target.wants/open-vm-tools.service → /lib/systemd/system/open-vm-tools.service.
Created symlink /etc/systemd/system/multi-user.target.wants/intel-sgx-load-module.service → /lib/systemd/system/intel-sgx-load-module.service.
Created symlink /etc/systemd/system/multi-user.target.wants/acpid.service → /lib/systemd/system/acpid.service.
Created symlink /etc/systemd/system/sysinit.target.wants/mdadm-shutdown.service → /lib/systemd/system/mdadm-shutdown.service.
Unit /lib/systemd/system/lvm2.service is masked, ignoring.
Unit /lib/systemd/system/mountdevsubfs.service is masked, ignoring.
Unit /lib/systemd/system/sudo.service is masked, ignoring.
Unit /lib/systemd/system/stop-bootlogd.service is masked, ignoring.
Unit /lib/systemd/system/bootmisc.service is masked, ignoring.
Unit /lib/systemd/system/checkfs.service is masked, ignoring.
Unit /lib/systemd/system/rmnologin.service is masked, ignoring.
Unit /lib/systemd/system/single.service is masked, ignoring.
Unit /lib/systemd/system/umountfs.service is masked, ignoring.
Unit /lib/systemd/system/umountnfs.service is masked, ignoring.
Unit /lib/systemd/system/motd.service is masked, ignoring.
Unit /lib/systemd/system/snapd.seeded.service is added as a dependency to a non-existent unit cloud-final.service.
Unit /lib/systemd/system/screen-cleanup.service is masked, ignoring.
Created symlink /etc/systemd/system/sysinit.target.wants/systemd-sysext.service → /lib/systemd/system/systemd-sysext.service.
Created symlink /etc/systemd/system/dbus-org.freedesktop.timesync1.service → /lib/systemd/system/systemd-timesyncd.service.
Created symlink /etc/systemd/system/sockets.target.wants/ssh.socket → /lib/systemd/system/ssh.socket.
Created symlink /etc/systemd/system/sysinit.target.wants/iscsid.service → /lib/systemd/system/iscsid.service.
Unit /lib/systemd/system/checkroot.service is masked, ignoring.
Created symlink /etc/systemd/system/network-online.target.wants/systemd-networkd-wait-online.service → /lib/systemd/system/systemd-networkd-wait-online.service.
Unit /lib/systemd/system/umountroot.service is masked, ignoring.
Unit /lib/systemd/system/x11-common.service is masked, ignoring.
root@ubuntu1804:~# while systemctl daemon-reload; do :; done

Broadcast message from systemd-journald@ubuntu1804.localdomain (Thu 2021-05-13 17:51:26 UTC):

systemd[1]: Caught <ABRT>, dumped core as pid 14176.


Broadcast message from systemd-journald@ubuntu1804.localdomain (Thu 2021-05-13 17:51:26 UTC):

systemd[1]: Freezing execution.

Failed to reload daemon: Connection reset by peer
```
</details>

<details>
<summary>Fedora Rawhide</summary>

```
[root@rawhide ~]# while systemctl daemon-reload; do :; done
^C
[root@rawhide ~]# systemctl preset-all
Removed /etc/systemd/system/nfs-idmapd.service.requires/nfs-convert.service.
Removed /etc/systemd/system/nfs-blkmap.service.requires/nfs-convert.service.
Removed /etc/systemd/system/rpc-statd-notify.service.requires/nfs-convert.service.
Removed /etc/systemd/system/cloud-init.target.wants/cloud-init.service.
Removed /etc/systemd/system/nfs-mountd.service.requires/nfs-convert.service.
Removed /etc/systemd/system/default.target.
Removed /etc/systemd/system/rpc-gssd.service.requires/nfs-convert.service.
Removed /etc/systemd/system/multi-user.target.wants/restraintd.service.
Removed /etc/systemd/system/multi-user.target.wants/nfs-client.target.
Removed /etc/systemd/system/rpc-statd.service.requires/nfs-convert.service.
Removed /etc/systemd/system/nfs-server.service.requires/nfs-convert.service.
Removed /etc/systemd/system/remote-fs.target.wants/nfs-client.target.
Created symlink /etc/systemd/system/dbus-org.fedoraproject.FirewallD1.service → /usr/lib/systemd/system/firewalld.service.
Created symlink /etc/systemd/system/multi-user.target.wants/firewalld.service → /usr/lib/systemd/system/firewalld.service.
Created symlink /etc/systemd/system/multi-user.target.wants/remote-cryptsetup.target → /usr/lib/systemd/system/remote-cryptsetup.target.
Created symlink /etc/systemd/system/multi-user.target.wants/machines.target → /usr/lib/systemd/system/machines.target.
Created symlink /etc/systemd/system/sysinit.target.wants/systemd-pstore.service → /usr/lib/systemd/system/systemd-pstore.service.
[root@rawhide ~]# while systemctl daemon-reload; do :; done

Broadcast message from systemd-journald@rawhide (Thu 2021-05-13 14:08:06 EDT):

systemd[1]: Caught <ABRT>, dumped core as pid 12574.

Failed to reload daemon: Connection reset by peer
[root@rawhide ~]# 
Broadcast message from systemd-journald@rawhide (Thu 2021-05-13 14:08:06 EDT):

systemd[1]: Freezing execution.

```
</details>

<details>
<summary>Fedora Rawhide - stack trace</summary>

```
May 13 14:08:05 rawhide systemd-rc-local-generator[12520]: /etc/rc.local is not marked executable, skipping.
May 13 14:08:05 rawhide systemd[1]: Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:660, function hashmap_iterate_in_internal_order(). Aborting.
May 13 14:08:05 rawhide audit[12574]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 subj=kernel pid=12574 comm="systemd" exe="/usr/lib/systemd/systemd" sig=6 res=1
May 13 14:08:05 rawhide systemd-coredump[12575]: Due to PID 1 having crashed coredump collection will now be turned off.
May 13 14:08:06 rawhide systemd-coredump[12575]: [🡕] Process 12574 (systemd) of user 0 dumped core.
                                                 
                                                 Found module linux-vdso.so.1 with build-id: 6ee388b3d5f40c155ab6024bf4c494dd5dc2a8a9
                                                 Found module libnl-3.so.200 with build-id: 4d5fd47dc316b7f4176965c4778a4ed6386fc68f
                                                 Found module libnl-route-3.so.200 with build-id: 1138966455ddcdb5159b150951d10411b1518997
                                                 Found module libibverbs.so.1 with build-id: 08dce64ef68a646ab25a67755e76fe03009a2b20
                                                 Found module libffi.so.6 with build-id: 7919f46637f314538a614abc03b9580b96b026d4
                                                 Found module libpcap.so.1 with build-id: fcc85ada80982959c669e46c87086ac39d1abd21
                                                 Found module libgpg-error.so.0 with build-id: 230316a110f9a8a589b2eceb1e634e56d4fd0722
                                                 Found module libattr.so.1 with build-id: e9fd1a86d3b4a76f70797e5abbc84ffbd7002d03
                                                 Found module libgcc_s.so.1 with build-id: c1f91d86bb8af7633815c33bb55ea7a455c80d2d
                                                 Found module libz.so.1 with build-id: f43e7898b82d9fda122bfbdf1a4859a3d832d987
                                                 Found module libcap-ng.so.0 with build-id: 4849cb55841ffe8ce1cbc913abe38ffbf22f222a
                                                 Found module libeconf.so.0 with build-id: 917f2967b9bf8596ff1d0d12bb6ab50caeecfe23
                                                 Found module libpcre2-8.so.0 with build-id: c6f202228469b4e4b3b7eb48782430034e9d1ebd
                                                 Found module ld-linux-x86-64.so.2 with build-id: a06d4bfcffa2015f0aa624a7ae4fcde286ccaf52
                                                 Found module libdl.so.2 with build-id: bd9524b011d76b7503c2f4c5be3248fddb78062a
                                                 Found module libm.so.6 with build-id: 90f3cec8091d5b5c4501b1883315b51e9c87ca50
                                                 Found module liblzma.so.5 with build-id: 6d4ebe2e0eeca0d174014bea3dfd56d6a816d263
                                                 Found module libzstd.so.1 with build-id: df6a3c9da745fcdcbe6311b001e2fcdf2277b78e
                                                 Found module libp11-kit.so.0 with build-id: 350d4280399a65731cad55f2384c4c75f090e27b
                                                 Found module libcrypto.so.1.1 with build-id: f913c4c3e57d32178701f736f3f877b3f548ca69
                                                 Found module liblz4.so.1 with build-id: a2429c266188acc10181f6936915f35274bb4a38
                                                 Found module libip4tc.so.2 with build-id: 66e2438196e5f96ab00b2607da51f1430c4485af
                                                 Found module libgcrypt.so.20 with build-id: 93035878149841ae8a7832d93d10256efeb25abb
                                                 Found module libcrypt.so.2 with build-id: 43277f1e080a002539733937400e2ddded479691
                                                 Found module libcap.so.2 with build-id: 4f71c26a42eada23856b3f5743644a1cf83d1b75
                                                 Found module libblkid.so.1 with build-id: 0220125b58ec151ea13bca7f3ca73f482cadfdfd
                                                 Found module libacl.so.1 with build-id: 492378f49e53bf592b736e56a67be45428879564
                                                 Found module libc.so.6 with build-id: b9d58e373350c522bc4bed251382a725e2acbe5a
                                                 Found module libpthread.so.0 with build-id: 77b96db9c393e36cbf046210421acf16eee17c14
                                                 Found module libkmod.so.2 with build-id: 3cc691315cb209dbb066ba18aab73ab11ed575be
                                                 Found module libaudit.so.1 with build-id: ca6584cc841c0d9f8182aad0c3edd0bd3b833d5f
                                                 Found module libpam.so.0 with build-id: 9bf2553d2e48c171e2f14738cd6e4a859458bf57
                                                 Found module libmount.so.1 with build-id: 62999d7d76ce8b0cab7ca1bec16bb287c1d24556
                                                 Found module libselinux.so.1 with build-id: 9ec3307b00e1cff9f5fb0b89806c62eb7a41c3f7
                                                 Found module libseccomp.so.2 with build-id: b68127d268bfcca1df3f180b752db98b30bac683
                                                 Found module librt.so.1 with build-id: a0006288806da6ef6aa9c5110c0624f17a53c214
                                                 Found module libsystemd-shared-248.so with build-id: 9de9436db751415e08f807ca29b38e0f862684bd
                                                 Found module systemd with build-id: 0000358d95249fbbd7a2d71370571ef3fb802fcc
                                                 Stack trace of thread 12574:
                                                 #0  0x00007f0248fa758b kill (libc.so.6 + 0x3d58b)
                                                 #1  0x000000000041cb2a crash (systemd + 0x1cb2a)
                                                 #2  0x00007f024914c4b0 __restore_rt (libpthread.so.0 + 0x134b0)
                                                 #3  0x00007f0248fa7292 raise (libc.so.6 + 0x3d292)
                                                 #4  0x00007f0248f908a4 abort (libc.so.6 + 0x268a4)
                                                 #5  0x00007f024942cb7e log_assert_failed (libsystemd-shared-248.so + 0x1d5b7e)
                                                 #6  0x00007f024941d25b hashmap_iterate_in_internal_order (libsystemd-shared-248.so + 0x1c625b)
                                                 #7  0x00007f024941d32d hashmap_iterate_entry (libsystemd-shared-248.so + 0x1c632d)
                                                 #8  0x00007f024941d35b _hashmap_iterate (libsystemd-shared-248.so + 0x1c635b)
                                                 #9  0x0000000000582736 hashmap_iterate (systemd + 0x182736)
                                                 #10 0x0000000000582bee target_add_default_dependencies (systemd + 0x182bee)
                                                 #11 0x0000000000582d4d target_load (systemd + 0x182d4d)
                                                 #12 0x000000000050d129 unit_load (systemd + 0x10d129)
                                                 #13 0x000000000049c86d manager_dispatch_load_queue (systemd + 0x9c86d)
                                                 #14 0x000000000049cd34 manager_load_unit (systemd + 0x9cd34)
                                                 #15 0x0000000000515b3d unit_add_dependency_by_name (systemd + 0x115b3d)
                                                 #16 0x0000000000568d95 device_add_udev_wants (systemd + 0x168d95)
                                                 #17 0x0000000000569e74 device_setup_unit (systemd + 0x169e74)
                                                 #18 0x0000000000569fc6 device_process_new (systemd + 0x169fc6)
                                                 #19 0x000000000056b4c4 device_enumerate (systemd + 0x16b4c4)
                                                 #20 0x000000000049a879 manager_enumerate (systemd + 0x9a879)
                                                 #21 0x00000000004a52a3 manager_reload (systemd + 0xa52a3)
                                                 #22 0x00000000004249d5 invoke_main_loop (systemd + 0x249d5)
                                                 #23 0x0000000000428091 main (systemd + 0x28091)
                                                 #24 0x00007f0248f91b75 __libc_start_main (libc.so.6 + 0x27b75)
                                                 #25 0x000000000041b92e _start (systemd + 0x1b92e)
May 13 14:08:06 rawhide systemd[1]: Caught <ABRT>, dumped core as pid 12574.
May 13 14:08:06 rawhide systemd[1]: Freezing execution.
May 13 14:08:06 rawhide systemctl[541]: Warning! D-Bus connection terminated.
May 13 14:08:06 rawhide systemctl[541]: Failed to wait for response: Connection reset by peer
May 13 14:08:06 rawhide systemd-oomd[503]: Failed to connect to /run/systemd/io.system.ManagedOOM: Connection refused
May 13 14:08:06 rawhide systemd-oomd[503]: Failed to acquire varlink connection: Connection refused
May 13 14:08:06 rawhide systemd-oomd[503]: Event loop failed: Connection refused
```
</details>

When you find out what's going on please let me know, because this is indeed quite strange.

poettering commented:
managed to reproduce the issue with that. thanks!

poettering commented:
So I think I nailed this one. Force pushed a new version. let's see if the CI agrees with me...

poettering commented:
As it appears I indeed managed to fix this one... ubuntu CI is slow these days, but at least the s390 one passed this time.

poettering commented:
Yippieh! All Green!

bluca commented on src/core/mount.c:
```diff
@@ -116,16 +134,33 @@ static bool mount_is_bind(const MountParameters *p) {
+        assert(m);
+
+        /* Determines whether to place a Requires= or BindsTo= dependency on the backing device unit. We do
+         * this by checking for the x-systemd.device-bound mount option. Iff it is set we use BindsTo=,
```
/Iff/if/

bluca replied:
TIL!

bluca commented on src/core/unit.c:
```diff
@@ -2728,7 +2707,36 @@ void unit_notify(Unit *u, UnitActiveState os, UnitActiveState ns, UnitNotifyFlag
 
-        unit_add_to_gc_queue(u);
+        /* And now, add the unit or depending units to various queues that will act on the new situation if
+         * needed. These queues generally check for continouus state changes rather than events (like most of
```
/continouus/continuous/

bluca commented on test/units/testsuite-57-short-lived.sh:
```diff
@@ -0,0 +1,18 @@
+set -ex
+
+if [ -f /tmp/testsuite-57.counter ] ; then
+    read counter < /tmp/testsuite-57.counter
```
shellcheck:

```
    read counter < /tmp/testsuite-57.counter
    ^--^ SC2162: read without -r will mangle backslashes.
```

bluca reviewed (approved):
LGTM - just a missing license header and some very minor nits. It is quite a large and complex change, the patches themselves and the explanations do make sense to me but I can't claim to know the existing state machine/queues logic too well. The new features are very very neat though!

poettering commented:
Force pushed a new version. Setting green label, since @bluca liked it and @keszybz already did one round.

fbuihuu commented:
@poettering someone recently reported a downstream issue which is essentially the same as the one described in https://github.com/systemd/systemd/issues/1741. That's the reason why I'm commenting only now.

IIUC `StopPropagatedFrom=` is supposed to address the aforementioned issue but I fail to see how exactly: as long as the dep on the wrong device is still included in the mount dependencies, the problem isn't solved, is it ?

The fact that the dep on the device is now edge-triggered doesn't help, it just moves the problem when the wrong device (the one PID1 initially tracked when the generated mount unit was loaded) is being stopped now.

What am I missing ?
